modules/diskq/diskq-serializer.c

```diff
@@ -45,6 +54,96 @@ _pread_all(int fd, char *buf,  size_t count, off_t offset)
   return retval + _pread_all(fd, buf+retval, count-retval, offset+retval);
 }
 
+static DiskQSerializerCache *
+_cache_new(void)
+{
+  DiskQSerializerCache *self = g_new0(DiskQSerializerCache, 1);
+  self->buffer = g_string_sized_new(CHUNK_SIZE);
+  self->pos = 0;
+
+  return self;
+}
+
+static void
+_cache_free(DiskQSerializerCache *self)
+{
+  g_string_free(self->buffer, TRUE);
+  g_free(self);
+}
+
+static gboolean
+_cache_fetch(DiskQSerializerCache *self, gchar *buf, gsize buflen)
+{
+  if (self->pos + buflen > self->buffer->len)
+    return FALSE;
+
+  memcpy(buf, &self->buffer->str[self->pos], buflen);
+  self->pos += buflen;
+  return TRUE;
+}
+
+static void
+_cache_move_data_to_front(DiskQSerializerCache *self)
+{
+  gsize data_len = self->buffer->len - self->pos;
+  memmove(self->buffer->str, &self->buffer->str[self->pos], data_len);
+  self->pos = 0;
+  g_string_set_size(self->buffer, data_len);
+}
+
+static gsize
+_cache_allocated_size(DiskQSerializerCache *self)
+{
+  return self->buffer->allocated_len;
+}
+
+static gsize
+_cache_data_size(DiskQSerializerCache *self)
+{
+  return (self->buffer->len - self->pos);
+}
+
+static void
+_cache_increase_buffer(DiskQSerializerCache *self, gsize buflen)
+{
+  gsize orig_len = self->buffer->len;
+  g_string_set_size(self->buffer, self->buffer->allocated_len + buflen);
+  g_string_set_size(self->buffer, orig_len);
+}
+
+static gchar *
+_cache_get_buffer_end(DiskQSerializerCache *self)
+{
+  return &self->buffer->str[self->buffer->len];
+}
+
+static void
+_cache_increase_data_size(DiskQSerializerCache *self, gsize buflen)
+{
+  g_string_set_size(self->buffer, self->buffer->len + buflen);
+}
+
+static gboolean
+_push_data_to_cache(MsgSerializer *self, gsize buflen)
+{
+  gsize chunk_size = MAX(buflen, CHUNK_SIZE);
+
+  if ((self->cache->pos % 16) == 0)
```

Comment:
Where does the `16` come from?

Reply:
This was accidental. Originally, I was fighting with some random corruption error problem, and I was suspecting it is related to alignment. This line supposed to avoid any alignment problem by only recycling the first part of the buffer, if the relevant data starts on a 16-divisible location. Moving the data in this case is safe. Later it turned out the failure came from the nventry check, so this is not necessary.